File: expansion-packs/bmad-research-dev/templates/servers/github/search-issues.ts
/**
 * GitHub Issues Search Server Wrapper
 * Wraps GitHub CLI (gh) and MCP GitHub tools
 *
 * Usage:
 *   import { searchIssues } from './servers/github/search-issues.ts'
 *   const issues = await searchIssues("label:bug is:open")
 */

interface Issue {
  number: number;
  title: string;
  state: 'open' | 'closed';
  author: string;
  labels: string[];
  assignees: string[];
  milestone?: {
    title: string;
    number: number;
  };
  body: string;
  createdAt: string;
  updatedAt: string;
  closedAt?: string;
  url: string;
  comments: number;
}

interface SearchOptions {
  state?: 'open' | 'closed' | 'all';
  labels?: string[];
  assignee?: string;
  milestone?: string;
  author?: string;
  limit?: number;
  sort?: 'created' | 'updated' | 'comments';
  order?: 'asc' | 'desc';
}

/**
 * Search GitHub issues in current repository
 *
 * @param query - GitHub search query (uses GitHub search syntax)
 * @param options - Optional search filters
 * @returns Array of issues matching the search
 *
 * @example
 * // Search open bugs
 * const bugs = await searchIssues("is:open label:bug")
 *
 * @example
 * // Search by milestone
 * const issues = await searchIssues("milestone:'Sprint 1'")
 *
 * @example
 * // Search with filters
 * const issues = await searchIssues("neural", {
 *   state: 'open',
 *   labels: ['type:experiment'],
 *   limit: 20
 * })
 */
export async function searchIssues(
  query?: string,
  options: SearchOptions = {}
): Promise<Issue[]> {
  console.log(`[GitHub] Searching issues: "${query || 'all'}"`)

  // Build GitHub CLI command
  let searchQuery = query || '';

  // Add filters
  if (options.state) {
    searchQuery += ` is:${options.state}`;
  }

  if (options.labels && options.labels.length > 0) {
    options.labels.forEach(label => {
      searchQuery += ` label:"${label}"`;
    });
  }

  if (options.assignee) {
    searchQuery += ` assignee:${options.assignee}`;
  }

  if (options.milestone) {
    searchQuery += ` milestone:"${options.milestone}"`;
  }

  if (options.author) {
    searchQuery += ` author:${options.author}`;
  }

  // Execute search using gh CLI
  const cmd = [
    'gh issue list',
    searchQuery ? `--search "${searchQuery.trim()}"` : '',
    options.limit ? `--limit ${options.limit}` : '--limit 50',
    '--json number,title,state,author,labels,assignees,milestone,body,createdAt,updatedAt,closedAt,url,comments'
  ].filter(Boolean).join(' ');

  console.log(`[GitHub] Executing: ${cmd}`)

  const result = await globalThis.Bash({ command: cmd });
  const issues = JSON.parse(result);

  console.log(`[GitHub] Found ${issues.length} issues`)

  return issues;
}

/**
 * Get all open issues
 */
export async function getOpenIssues(limit?: number): Promise<Issue[]> {
  return searchIssues(undefined, {
    state: 'open',
    limit: limit || 100
  });
}

/**
 * Get issues by label
 */
export async function getIssuesByLabel(label: string): Promise<Issue[]> {
  return searchIssues(undefined, {
    labels: [label],
    state: 'all'
  });
}

/**
 * Get issues in milestone
 */
export async function getIssuesByMilestone(milestone: string): Promise<Issue[]> {
  return searchIssues(undefined, {
    milestone,
    state: 'all'
  });
}

/**
 * Get issues assigned to user
 */
export async function getAssignedIssues(assignee: string): Promise<Issue[]> {
  return searchIssues(undefined, {
    assignee,
    state: 'open'
  });
}

/**
 * Get issues by type
 */
export async function getIssuesByType(type: string): Promise<Issue[]> {
  const typeLabels: Record<string, string> = {
    'epic': 'type:epic',
    'story': 'type:story',
    'task': 'type:task',
    'bug': 'type:bug',
    'experiment': 'type:experiment',
    'analysis': 'type:analysis'
  };

  const label = typeLabels[type.toLowerCase()] || `type:${type}`;

  return getIssuesByLabel(label);
}

/**
 * Get issues by status
 */
export async function getIssuesByStatus(status: string): Promise<Issue[]> {
  const statusLabels: Record<string, string> = {
    'backlog': 'status:backlog',
    'todo': 'status:todo',
    'doing': 'status:doing',
    'review': 'status:review',
    'done': 'status:done'
  };

  const label = statusLabels[status.toLowerCase()] || `status:${status}`;

  return getIssuesByLabel(label);
}

/**
 * Search research-related issues
 */
export async function searchResearchIssues(
  keywords?: string
): Promise<{
  experiments: Issue[];
  literature: Issue[];
  analysis: Issue[];
  papers: Issue[];
}> {
  const [experiments, literature, analysis, papers] = await Promise.all([
    getIssuesByLabel('research:experiment'),
    getIssuesByLabel('research:literature'),
    getIssuesByLabel('research:analysis'),
    getIssuesByLabel('research:paper')
  ]);

  // Filter by keywords if provided
  const filterByKeywords = (issues: Issue[]) => {
    if (!keywords) return issues;
    const lowerKeywords = keywords.toLowerCase();
    return issues.filter(issue =>
      issue.title.toLowerCase().includes(lowerKeywords) ||
      issue.body.toLowerCase().includes(lowerKeywords)
    );
  };

  return {
    experiments: filterByKeywords(experiments),
    literature: filterByKeywords(literature),
    analysis: filterByKeywords(analysis),
    papers: filterByKeywords(papers)
  };
}

/**
 * Get issue statistics
 */
export async function getIssueStats(): Promise<{
  total: number;
  open: number;
  closed: number;
  byLabel: Record<string, number>;
  byMilestone: Record<string, number>;
  byAssignee: Record<string, number>;
  avgTimeToClose: number; // days
}> {
  const allIssues = await searchIssues(undefined, {
    state: 'all',
    limit: 1000
  });

  const open = allIssues.filter(i => i.state === 'open').length;
  const closed = allIssues.filter(i => i.state === 'closed').length;

  // Count by label
  const byLabel: Record<string, number> = {};
  allIssues.forEach(issue => {
    issue.labels.forEach(label => {
      byLabel[label] = (byLabel[label] || 0) + 1;
    });
  });

  // Count by milestone
  const byMilestone: Record<string, number> = {};
  allIssues.forEach(issue => {
    if (issue.milestone) {
      const title = issue.milestone.title;
      byMilestone[title] = (byMilestone[title] || 0) + 1;
    }
  });

  // Count by assignee
  const byAssignee: Record<string, number> = {};
  allIssues.forEach(issue => {
    issue.assignees.forEach(assignee => {
      byAssignee[assignee] = (byAssignee[assignee] || 0) + 1;
    });
  });

  // Calculate average time to close
  const closedIssues = allIssues.filter(i => i.closedAt);
  const avgTimeToClose = closedIssues.length > 0
    ? closedIssues.reduce((sum, issue) => {
        const created = new Date(issue.createdAt).getTime();
        const closed = new Date(issue.closedAt!).getTime();
        return sum + (closed - created);
      }, 0) / closedIssues.length / (1000 * 60 * 60 * 24) // Convert to days
    : 0;

  return {
    total: allIssues.length,
    open,
    closed,
    byLabel,
    byMilestone,
    byAssignee,
    avgTimeToClose
  };
}

/**
 * Find related issues
 */
export async function findRelatedIssues(
  issueNumber: number
): Promise<Issue[]> {
  // Get the issue details
  const cmd = `gh issue view ${issueNumber} --json number,title,body,labels`;
  const result = await globalThis.Bash({ command: cmd });
  const issue = JSON.parse(result);

  // Extract keywords from title
  const keywords = issue.title
    .toLowerCase()
    .split(/\s+/)
    .filter((word: string) => word.length > 4)
    .slice(0, 3)
    .join(' ');

  // Search for similar issues
  const related = await searchIssues(keywords, {
    state: 'all',
    limit: 10
  });

  // Filter out the original issue
  return related.filter(i => i.number !== issueNumber);
}

File: expansion-packs/bmad-research-dev/templates/servers/github/search_issues.py
"""
GitHub Issues Search Server Wrapper
Wraps GitHub CLI (gh) and MCP GitHub tools

Usage:
    from servers.github.search_issues import search_issues
    issues = search_issues("label:bug is:open")
"""

import json
import subprocess
from typing import List, Dict, Optional, Literal
from datetime import datetime


def search_issues(
    query: Optional[str] = None,
    state: Optional[Literal['open', 'closed', 'all']] = None,
    labels: Optional[List[str]] = None,
    assignee: Optional[str] = None,
    milestone: Optional[str] = None,
    author: Optional[str] = None,
    limit: int = 50,
    sort: Optional[Literal['created', 'updated', 'comments']] = None,
    order: Optional[Literal['asc', 'desc']] = None
) -> List[Dict]:
    """
    Search GitHub issues in current repository

    Args:
        query: GitHub search query (uses GitHub search syntax)
        state: Issue state filter
        labels: Label filters
        assignee: Assignee filter
        milestone: Milestone filter
        author: Author filter
        limit: Maximum results (default: 50)
        sort: Sort field
        order: Sort order

    Returns:
        Array of issues matching the search

    Examples:
        # Search open bugs
        bugs = search_issues("is:open label:bug")

        # Search by milestone
        issues = search_issues("milestone:'Sprint 1'")

        # Search with filters
        issues = search_issues("neural", state='open', labels=['type:experiment'], limit=20)
    """
    print(f"[GitHub] Searching issues: \"{query or 'all'}\"")

    # Build search query
    search_query = query or ''

    # Add filters
    if state:
        search_query += f" is:{state}"

    if labels:
        for label in labels:
            search_query += f" label:\"{label}\""

    if assignee:
        search_query += f" assignee:{assignee}"

    if milestone:
        search_query += f" milestone:\"{milestone}\""

    if author:
        search_query += f" author:{author}"

    # Build command
    cmd = ['gh', 'issue', 'list']

    if search_query.strip():
        cmd.extend(['--search', search_query.strip()])

    cmd.extend(['--limit', str(limit)])
    cmd.extend([
        '--json',
        'number,title,state,author,labels,assignees,milestone,body,createdAt,updatedAt,closedAt,url,comments'
    ])

    print(f"[GitHub] Executing: {' '.join(cmd)}")

    result = subprocess.run(cmd, capture_output=True, text=True, check=True)
    issues = json.loads(result.stdout)

    print(f"[GitHub] Found {len(issues)} issues")

    return issues


def get_open_issues(limit: int = 100) -> List[Dict]:
    """Get all open issues"""
    return search_issues(state='open', limit=limit)


def get_issues_by_label(label: str) -> List[Dict]:
    """Get issues by label"""
    return search_issues(labels=[label], state='all')


def get_issues_by_milestone(milestone: str) -> List[Dict]:
    """Get issues in milestone"""
    return search_issues(milestone=milestone, state='all')


def get_assigned_issues(assignee: str) -> List[Dict]:
    """Get issues assigned to user"""
    return search_issues(assignee=assignee, state='open')


def get_issues_by_type(issue_type: str) -> List[Dict]:
    """Get issues by type"""
    type_labels = {
        'epic': 'type:epic',
        'story': 'type:story',
        'task': 'type:task',
        'bug': 'type:bug',
        'experiment': 'type:experiment',
        'analysis': 'type:analysis'
    }

    label = type_labels.get(issue_type.lower(), f"type:{issue_type}")

    return get_issues_by_label(label)


def get_issues_by_status(status: str) -> List[Dict]:
    """Get issues by status"""
    status_labels = {
        'backlog': 'status:backlog',
        'todo': 'status:todo',
        'doing': 'status:doing',
        'review': 'status:review',
        'done': 'status:done'
    }

    label = status_labels.get(status.lower(), f"status:{status}")

    return get_issues_by_label(label)


def search_research_issues(keywords: Optional[str] = None) -> Dict[str, List[Dict]]:
    """Search research-related issues"""
    experiments = get_issues_by_label('research:experiment')
    literature = get_issues_by_label('research:literature')
    analysis = get_issues_by_label('research:analysis')
    papers = get_issues_by_label('research:paper')

    # Filter by keywords if provided
    def filter_by_keywords(issues: List[Dict]) -> List[Dict]:
        if not keywords:
            return issues
        lower_keywords = keywords.lower()
        return [
            issue for issue in issues
            if lower_keywords in issue['title'].lower() or
               lower_keywords in issue.get('body', '').lower()
        ]

    return {
        'experiments': filter_by_keywords(experiments),
        'literature': filter_by_keywords(literature),
        'analysis': filter_by_keywords(analysis),
        'papers': filter_by_keywords(papers)
    }


def get_issue_stats() -> Dict:
    """Get issue statistics"""
    all_issues = search_issues(state='all', limit=1000)

    open_count = len([i for i in all_issues if i['state'] == 'open'])
    closed_count = len([i for i in all_issues if i['state'] == 'closed'])

    # Count by label
    by_label: Dict[str, int] = {}
    for issue in all_issues:
        for label in issue.get('labels', []):
            label_name = label.get('name', '') if isinstance(label, dict) else str(label)
            by_label[label_name] = by_label.get(label_name, 0) + 1

    # Count by milestone
    by_milestone: Dict[str, int] = {}
    for issue in all_issues:
        milestone = issue.get('milestone')
        if milestone:
            title = milestone.get('title', '') if isinstance(milestone, dict) else str(milestone)
            by_milestone[title] = by_milestone.get(title, 0) + 1

    # Count by assignee
    by_assignee: Dict[str, int] = {}
    for issue in all_issues:
        for assignee in issue.get('assignees', []):
            assignee_name = assignee.get('login', '') if isinstance(assignee, dict) else str(assignee)
            by_assignee[assignee_name] = by_assignee.get(assignee_name, 0) + 1

    # Calculate average time to close
    closed_issues = [i for i in all_issues if i.get('closedAt')]
    if closed_issues:
        total_time = 0
        for issue in closed_issues:
            created = datetime.fromisoformat(issue['createdAt'].replace('Z', '+00:00'))
            closed = datetime.fromisoformat(issue['closedAt'].replace('Z', '+00:00'))
            total_time += (closed - created).total_seconds()
        avg_time_to_close = (total_time / len(closed_issues)) / (60 * 60 * 24)  # Convert to days
    else:
        avg_time_to_close = 0

    return {
        'total': len(all_issues),
        'open': open_count,
        'closed': closed_count,
        'byLabel': by_label,
        'byMilestone': by_milestone,
        'byAssignee': by_assignee,
        'avgTimeToClose': avg_time_to_close
    }


def find_related_issues(issue_number: int) -> List[Dict]:
    """Find related issues"""
    # Get the issue details
    cmd = ['gh', 'issue', 'view', str(issue_number), '--json', 'number,title,body,labels']
    result = subprocess.run(cmd, capture_output=True, text=True, check=True)
    issue = json.loads(result.stdout)

    # Extract keywords from title
    keywords = [
        word for word in issue['title'].lower().split()
        if len(word) > 4
    ][:3]

    search_query = ' '.join(keywords)

    # Search for similar issues
    related = search_issues(search_query, state='all', limit=10)

    # Filter out the original issue
    return [i for i in related if i['number'] != issue_number]
